src/plonk/lookup/prover.rs

```diff
@@ -0,0 +1,652 @@
+        // where a'(X) is the compression of the permuted input columns,
+        // s_j(X) is the jth table column in this lookup,
+        // s'(X) is the compression of the permuted table columns,
+        // and i is the ith row of the column.
+        let mut lookup_product = vec![C::Scalar::zero(); params.n as usize];
+        // Denominator uses the permuted input column and permuted table column
+        parallelize(&mut lookup_product, |lookup_product, start| {
+            for ((lookup_product, permuted_input_value), permuted_table_value) in lookup_product
+                .iter_mut()
+                .zip(self.permuted_input_column[start..].iter())
+                .zip(self.permuted_table_column[start..].iter())
+            {
+                *lookup_product = (*beta + permuted_input_value) * &(*gamma + permuted_table_value);
+            }
+        });
+
+        // Batch invert to obtain the denominators for the lookup product
+        // polynomials
+        lookup_product.iter_mut().batch_invert();
+
+        // Finish the computation of the entire fraction by computing the numerators
+        // (\theta^{m-1} a_1(\omega^i) + \theta^{m-2} a_2(\omega^i) + ... + \theta a_{m-1}(\omega^i) + a_m(\omega^i) + \beta)(\theta^{m-1} s_1(\omega^i) + \theta^{m-2} s_2(\omega^i) + ... + \theta s_{m-1}(\omega^i) + s_m(\omega^i) + \gamma)
+        // Compress unpermuted input columns
+        let mut input_term = vec![C::Scalar::zero(); params.n as usize];
+        for unpermuted_input_column in self.unpermuted_input_columns.iter() {
+            parallelize(&mut input_term, |input_term, start| {
+                for (input_term, input_value) in input_term
+                    .iter_mut()
+                    .zip(unpermuted_input_column[start..].iter())
+                {
+                    *input_term *= &theta;
+                    *input_term += input_value;
+                }
+            });
+        }
+
+        // Compress unpermuted table columns
+        let mut table_term = vec![C::Scalar::zero(); params.n as usize];
+        for unpermuted_table_columns in self.unpermuted_table_columns.iter() {
+            parallelize(&mut table_term, |table_term, start| {
+                for (table_term, fixed_value) in table_term
+                    .iter_mut()
+                    .zip(unpermuted_table_columns[start..].iter())
+                {
+                    *table_term *= &theta;
+                    *table_term += fixed_value;
+                }
+            });
+        }
+
+        // Add \beta and \gamma offsets
+        parallelize(&mut lookup_product, |product, start| {
+            for ((product, input_term), table_term) in product
+                .iter_mut()
+                .zip(input_term[start..].iter())
+                .zip(table_term[start..].iter())
+            {
+                *product *= &(*input_term + &beta);
+                *product *= &(*table_term + &gamma);
+            }
+        });
```

Comment:
These three loops could be integrated together to avoid the extra allocations and choke points for parallelism, and also make it more clear what's happening.

```suggestion
        // Finish the computation of the entire fraction by computing the numerators
        // (\theta^{m-1} a_1(\omega^i) + \theta^{m-2} a_2(\omega^i) + ... + \theta a_{m-1}(\omega^i) + a_m(\omega^i) + \beta)
        // * (\theta^{m-1} s_1(\omega^i) + \theta^{m-2} s_2(\omega^i) + ... + \theta s_{m-1}(\omega^i) + s_m(\omega^i) + \gamma)
        parallelize(&mut lookup_product, |product, start| {
            for (i, product) in product.iter_mut().enumerate() {
                let i = i + start;

                // Compress unpermuted input columns
                let mut input_term = C::Scalar::zero();
                for unpermuted_input_column in self.unpermuted_input_columns.iter() {
                    input_term *= &theta;
                    input_term += &unpermuted_input_column[i];
                }

                // Compress unpermuted table columns
                let mut table_term = C::Scalar::zero();
                for unpermuted_table_column in self.unpermuted_table_columns.iter() {
                    table_term *= &theta;
                    table_term += &unpermuted_table_column[i];
                }

                *product *= &(input_term + &beta);
                *product *= &(table_term + &gamma);
            }
        });
```